evennia/contrib/mapbuilder.py

```diff
@@ -0,0 +1,256 @@
+
+This is a command capable of taking a reference to a basic 2D ASCII
+map stored in the Evennia directory and  generating the rooms and exits 
+necessary to populate that world. The characters of the map are iterated
+over and compared to a list of trigger characters. When a match is found
+it triggers the corresponding instructions. Use by importing and including 
+the command in your default_cmdsets module. For example:
+
+    # mygame/commands/default_cmdsets.py
+    
+    from evennia.contrib import mapbuilder
+    
+    ...
+    
+    self.add(mapbuilder.CmdMapBuilder())
+
+You then call the command in-game using the path to the module and the
+name of the variable holding the map string. The path you provide is
+relative to the evennia or your mygame folder.
+
+    @mapbuilder <path.to.module.VARNAME>
+
+For example to generate from the sample map in this module:
+
+    @mapbuilder evennia.contrib.mapbuilder.EXAMPLE_MAP
+
+Whilst this map is contained here for convenience, it is suggested
+that your map be stored in a separate stand alone module in the
+mygame/world folder accessed with @mapbuilder world.gamemap.MAP
+
+The rooms generated for each square of the map are generated using 
+instructions for each room type. These instructions are intended 
+to be changed and adapted for your purposes. In writing these 
+instructions you have access to the full API just like Batchcode.
+
+"""
+# ---------- #
+
+# This code should be in mymap.py in mygame/world.
+
+# -*- coding: utf-8 -*-
+
+EXAMPLE_MAP = """\
+≈≈≈≈≈≈≈≈≈
+≈≈≈≈≈≈≈≈≈
+≈≈♣♠♣♠♣≈≈
+≈≈♠n∩n♠≈≈
+≈≈♣∩▲∩♣≈≈
+≈≈♠n≈n♠≈≈
+≈≈♣♠≈♠♣≈≈
+≈≈≈≈≈≈≈≈≈
+≈≈≈≈≈≈≈≈≈
+"""
+
+# ---------- #
+
+#Add the necessary imports for your instructions here.
+from evennia import create_object
+from typeclasses import rooms, exits
+from evennia.utils import utils
+from random import randint
+import random
+
+def build_map(caller, raw_map):
+    """
+    This is the part of the code designed to be changed and expanded. It
+    contains the instructions that are called when a match is made between
+    the characters in your map and the characters that trigger the building
+    instructions.
+    
+    """
+    
+    #Create a tuple containing the trigger characters
+    forest = ("♣", "♠")
+    #Create a function that contains the instructions.
+    def create_forest(x, y):
+        #This has just basic instructions, building and naming the room.
+        room = create_object(rooms.Room, key="forest" + str(x) + str(y))
+        room.db.desc = "Basic forest room."
+        
+        #Always include this at the end. Sets up for advanced functions.
+        caller.msg(room.key + " " + room.dbref)
+        room_list.append([room, x, y])
+ 
+ 
+    mountains = ("∩", "n")
+    def create_mountains(x, y):
+        #We'll do something fancier in this one.
+        room = create_object(rooms.Room, key="mountains" + str(x) + str(y))
+        
+        #We'll select a description at random from a list.
+        room_desc = [
```

Comment:
Since this function appears to actually be meant to supply hard-coded data, maybe the instructions for building each of these terrain types could be moved outside of the mapbuilder code and into something equivalent to how evmenu nodes work? 
Here's a suggested alternative: Make the call `@mapbuilder <mapfile> [legendfile]`, where legendfile defaults to the same file as the map file. The legend file defines a "legend dict" that can either have a fixed value or one you give as part of the path. The legend dict maps map symbols to functions in that same module. Here's an example:

```python
# file map_legend.py 

# map might optionally be up here^

MAP_LEGEND = \
    {"▲": build_temple, 
     "∩": build_mountain,
     "n": build_mountain}

def build_mountain(x, y):
    # build mountain
    return mountain

def build_temple(x, y):
    # build temple
    return temple
```

This would abstract the build mechanism from your map-builder function completely - for every position on the grid, the system queries `MAP_LEGEND` and calls the linked function ...